src/MMALSharp.Processing/Processors/Effects/ConvolutionBase.cs

```diff
@@ -27,165 +52,76 @@ public abstract class ConvolutionBase
-            IntPtr pNative = IntPtr.Zero;
-            int bytes;
-            byte[] store = null;
-
-            using (var ms = new MemoryStream(context.Data))
-            using (var bmp = this.LoadBitmap(context, ms))
+            if (!context.Raw)
             {
-                bmpData = bmp.LockBits(new Rectangle(0, 0,
-                        bmp.Width,
-                        bmp.Height),
-                    ImageLockMode.ReadWrite,
-                    bmp.PixelFormat);
-
-                if (context.Raw)
-                {
-                    this.InitBitmapData(context, bmpData);
-                }
-
-                pNative = bmpData.Scan0;
-
-                // Split image into 4 quadrants and process individually.
-                var quadA = new Rectangle(0, 0, bmpData.Width / 2, bmpData.Height / 2);
-                var quadB = new Rectangle(bmpData.Width / 2, 0, bmpData.Width / 2, bmpData.Height / 2);
-                var quadC = new Rectangle(0, bmpData.Height / 2, bmpData.Width / 2, bmpData.Height / 2);
-                var quadD = new Rectangle(bmpData.Width / 2, bmpData.Height / 2, bmpData.Width / 2, bmpData.Height / 2);
-
-                bytes = bmpData.Stride * bmp.Height;
-
-                var rgbValues = new byte[bytes];
-
-                // Copy the RGB values into the array.
-                Marshal.Copy(pNative, rgbValues, 0, bytes);
-
-                var bpp = Image.GetPixelFormatSize(bmp.PixelFormat) / 8;
-
-                var t1 = Task.Run(() =>
-                {
-                    this.ProcessQuadrant(quadA, bmp, bmpData, rgbValues, kernel, kernelWidth, kernelHeight, bpp);
-                });
-                var t2 = Task.Run(() =>
-                {
-                    this.ProcessQuadrant(quadB, bmp, bmpData, rgbValues, kernel, kernelWidth, kernelHeight, bpp);
-                });
-                var t3 = Task.Run(() =>
-                {
-                    this.ProcessQuadrant(quadC, bmp, bmpData, rgbValues, kernel, kernelWidth, kernelHeight, bpp);
-                });
-                var t4 = Task.Run(() =>
-                {
-                    this.ProcessQuadrant(quadD, bmp, bmpData, rgbValues, kernel, kernelWidth, kernelHeight, bpp);
-                });
-
-                Task.WaitAll(t1, t2, t3, t4);
-
-                if (context.Raw)
-                {
-                    store = new byte[bytes];
-                    Marshal.Copy(pNative, store, 0, bytes);
-                }
-
-                bmp.UnlockBits(bmpData);
-
-                if (!context.Raw)
-                {
-                    using (var ms2 = new MemoryStream())
-                    {
-                        bmp.Save(ms2, context.StoreFormat);
-                        store = new byte[ms2.Length];
-                        Array.Copy(ms2.ToArray(), 0, store, 0, ms2.Length);
-                    }
-                }
+                throw new Exception("Convolution effects require raw frame data");
```

Comment:
This concerns me a bit. I'm quite sure that image convolution worked on encoded images before - if they were encoded they were loaded into a new `Bitmap` instance and the metadata was made available via `BitmapData`. Is that not going to be possible in this re-write? Ideally it should be able to handle both.

Reply:
I think you're right, I think an optional extra step to load it into a new bitmap should work. I'll take a look at it from that perspective. It does ultimately need to be raw, though, so that `FrameAnalyser` can derive the bits (and bytes) per pixel. I don't think the in-memory bitmap conversion is too expensive. (I sort of wish we had a good abstraction for turning performance data on and off... maybe something for the future.)

Speaking of bytes per pixel, I'm still struggling to get `TakeRawPicture` to work. When I try the code shown below, `ImageContext.PixelFormat` is not set which causes `FrameAnalyser` to fail in the `GetBytesPerPixel` method. Am I overlooking something?

```csharp
var cam = GetConfiguredCamera();
MMALCameraConfig.Encoding = MMALEncoding.RGB24;
MMALCameraConfig.EncodingSubFormat = MMALEncoding.RGB24;
cam.ConfigureCameraSettings();
using (var imgCaptureHandler = new ImageStreamCaptureHandler(ramdiskPath, "jpg"))
{
    imgCaptureHandler.Manipulate(context =>
    {
        context.Apply(new SharpenProcessor());
    }, ImageFormat.Jpeg);
    await cam.TakeRawPicture(imgCaptureHandler);
}
cam.Cleanup();
```

Ah -- looking at the debug log, it looks like `StreamCaptureHandler` is where the problem may originate. In the log I see `Applying image processing` followed by `Something went wrong while processing stream` and the null ref throw. (It was initially confusing because the handler only logs the inner exception -- there was none -- and not the actual exception!)

Reply:
Let me run my thinking by you:

Since my revisions operate on `ImageContext` data, do you know if copying data out of a _formatted_ `Bitmap` (sourced from a JPEG memory stream for example) yields pixel level data? I was trying to figure out if I need to put it through a format conversion to get raw data. The old version operated directly on the locked bitmap, I'm wondering how much voodoo the `Bitmap` object does to make the pixels of a lossy compressed image available (once you unlock it, I assume it's back to lossy JPEG again). I know when a bitmap region is locked, .NET actually makes another copy of that region and pins it in memory, so I'm guessing that it probably is decompressed at that point. So I suspect copying that really will get uncompressed pixels.

Originally I was thinking after the context is updated (by copying the locked `Bitmap` data), the point above implies `ImageContext.Raw` should then be `true` -- but then I realized that's probably not correct. The raw flag only applies to the three RGB formats, and I suppose the copied bitmap could be something like YUV, does that sound right?

So if that's right, after the frame is processed, the raw-to-formatted conversion really shouldn't necessarily require raw input at all, either. 

Do I sound like I'm on the right track here?

Reply:
I was under the impression that the compression algorithm JPEG and other lossy formats use would remove full pixels, meaning the R, G, B and A if applicable would be removed. This means that you can still iterate over the data in the same way you would with raw data but there just wouldn't be as much of it. Once the data is lost by the lossy algorithm, the `Bitmap` class would have no way of retrieving that data again. Hopefully I'm right in my assumptions and apologies for taking so long to get back to you on this.

Reply:
After doing a bunch of reading and more thinking about it, my question doesn't really make sense. JPEG is lossy but it isn't storing pixels nor is it "losing" pixels, what it's losing is more abstract (DCT compression is based on perceptual color models and frequency thresholds). So it would never really make sense to operate on the compressed version of an image, the only thing that data is useful for is to decompress back to a bitmap -- it isn't stored as pixels, in other words, it's just a data stream meant to feed into a decompression algorithm.

I was basically mixing up "raw" with "uncompressed" -- the "raw" formats are image data which has never been compressed, versus "uncompressed" meaning a full-pixel-count bitmap that was _stored_ as compressed. We'll have "all" of the pixels, but they will simply exhibit compression artifacts here and there.

I'm still hung up on `TakeRawPicture` not passing a correct `ImageContext`, even though I will get it working with non-raw images as well. I assume a raw image will perform better without the overhead of bitmap locking and copying.

I also have to apologize for taking so long to deal with this. Work is really running me ragged right now -- I'm doing weekends, evenings, etc. (spent about 20 hours last weekend starting [this](https://github.com/MV10/EventStreamDotNet/)).

Reply:
Thanks for the clarification, I should have done more research before replying. Are you suggesting that the `Bitmap` class does in fact decompress the data when it gets initialised? Not sure why you're still having trouble, my changes in 8b07ab3 should have set the pixel format [here](https://github.com/techyian/MMALSharp/blob/8b07ab3b697ff1159d988db9a84e0501a7b331e6/src/MMALSharp/Components/MMALCameraComponent.cs#L340). I'll try and have a look today/tomorrow.

Reply:
> Are you suggesting that the Bitmap class does in fact decompress the data when it gets initialised?

Exactly. When I said this:

> So it would never really make sense to operate on the compressed version of an image, the only thing that data is useful for is to decompress back to a bitmap -- it isn't stored as pixels, in other words, it's just a data stream meant to feed into a decompression algorithm.

I was essentially referring to the contents of a full `ImageContext` frame buffer when MMAL is outputting compressed data. I don't know what `Bitmap` does internally, maybe it's more efficient to store it compressed, but that doesn't matter to us. Once we lock it for read/write, the pinned copy is definitely decompressed to full resolution. The loss simply refers to image quality, not actual individual pixels.

About the `TakeRawPicture` problem and `ImageContext.PixelFormat` -- using the code shown above, and adding one line of console output to the sharpen `Apply` method outputs `PixelFormat is null: True`:

```csharp
public void Apply(ImageContext context)
{
    Console.WriteLine($"PixelFormat is null: {context.PixelFormat == null}");
    this.ApplyConvolution(_kernel, KernelWidth, KernelHeight, context);
}
```

Your change was 24 days ago and I caught up my branch 20 days ago, so I don't think my branch is behind now.

Reply:
Ooooh... I forgot about another weird behavior I was seeing with `TakeRawPicture` -- it seems to terminate the program before the convolution processing actually finishes running. I mentioned that [here](https://github.com/techyian/MMALSharp/issues/108#issuecomment-689754922) after I added a bunch of output to trace line-by-line:

> I take that back. It's calling the delegate but it never completes. I guess that call is on another thread and the app shuts down before `ApplyConvolution` runs more than a couple of lines.

Reply:
It definitely only gets called once for me, I'm using your branch with the latest changes. I did commit a change which prevents the Callback Handler from being called if the buffer size is empty which prevents `Manipulate` from being called twice. Not sure why you're still seeing that, are you modifying the resolution at all or just leaving everything except the encoding & pixel formats the same?

> Ooooh... I forgot about another weird behavior I was seeing with TakeRawPicture -- it seems to terminate the program before the convolution processing actually finishes running. I mentioned that here after I added a bunch of output to trace line-by-line:

Again, not something I can reproduce. I've put a log call at the end of the `ApplyConvolution` method and it's outputting to the log OK. Am I missing something here?

As an aside, I'm still seeing a blue hue when applying convolutions to a raw image and then saving the output to an encoded format via `Bitmap.Save()`. If I run the raw modified image through a `MMALImageEncoder` instead of allowing `Bitmap` to save then the colours are OK but there are other issues that occur when I do this which I'll make separate tickets for. None of this is a regression based on your work from what I can see, although v0.6 appeared to support applying convolutions to raw images, this in fact doesn't appear to be the case and only ever worked with encoded formats.

Reply:
I took Friday off from work, so fingers crossed, I'm going to tackle this bright and early tomorrow!

Reply:
Here's a weird finding to start the day:

My `TakeRawPicture` test set the camera to 1296x972 as a side-effect of earlier motion-detection testing (it's a full-frame binning mode). To my surprise, a different resolution was reported in an exception in the debug log: `Resolution 1312x976 has no recommended cell counts.` After a bit of searching, it turns out that the driver bumps the buffer up to the nearest 32-byte value ([RPi forum discussion](https://www.raspberrypi.org/forums/viewtopic.php?t=62364&start=275#p522410)):

> Because of the way the HW works, the width needs to be expanded to the nearest 32 bytes. So 1296 gets rounded up to 1312. The amount of valid data is still the same.

Have you encountered this before? Given the "valid data" comment, I'm thinking the `Resolution` field in `ImageContext` is a bit of a misnomer. It sounds as if only the width is padded to 32 bytes, so perhaps what's needed is a new field, `BufferWidth`, and to somehow set `Resolution.Width` to the true horizontal pixel count (which we still need for correct processing).

I can work around this for now just by testing with a resolution naturally aligned to 32 bytes.

(The exception is also the reason I was saying earlier that execution was somehow being interrupted: I had expected an exception to be reported in the console by my outer `Program.Main` catch-block then terminate the program, but `StreamCaptureHandler` catches it and logs it -- which means it fails silently if logging is disabled . This design might be something to reconsider since this had unexpected side-effects.)

Reply:
I hope I don't bother you too much with my thinking-out-loud style of working through this stuff. 💡 

The output port's zero-length buffer check doesn't seem to work consistently. (My copy definitely has that change now.)

Even though it's writing a processed JPG file now (albeit a blue one!), in the debug log I saw an index out of range error from the parallel cell processing loop, so I disabled that and wrote a serial loop with logging, thinking I'd see some kind of off-by-1 problem. Instead, the still image frame is processed, then a zero-length buffer is passed to the algorithm during shutdown, and that's where it fails. I assume the exception is some sort of dealloc/processing race condition, but that aspect is a side-effect -- it shouldn't be processing a zero-length buffer in the first place. It looks like that third output buffer (you mentioned earlier that it allocates three) _does_ correctly skip processing which is even more puzzling.

I'm wondering if you can think of where in the pipeline this might go wrong.

The first line of the log snippet below is the end of the correct cell processing (now at 640 x 480, hence 1024 cells). Then you can see a zero-length EOS buffer, and you can see that image processing kicks off again. The last entry below is the first pass of another cell processing loop. (It gets to cell 64 then throws index out of range, but obviously it shouldn't even start processing.)

I was thinking I'd short-circuit the whole thing on zero-length further downstream (I'm thinking in `PostProcess`), then I realized that has a zero-length-stream check, so I logged that. Towards the end of the logs below, you can see that the stream has 6K of "something" in it, which is why I think this second invocation happens. Not sure how to reconcile that with the zero buffer length though.

```
2020-10-09 08:33:01.369 -04:00 [INF] cell index 1023
2020-10-09 08:33:01.468 -04:00 [DBG] Releasing buffer.
2020-10-09 08:33:01.469 -04:00 [DBG] vc.ril.camera:out:2(RGB3): End of stream. Signaling completion...
2020-10-09 08:33:01.472 -04:00 [DBG] vc.ril.camera:out:2(RGB3): In native StillPort output callback
2020-10-09 08:33:01.472 -04:00 [DBG] vc.ril.camera:out:2(RGB3): In native output callback
2020-10-09 08:33:01.473 -04:00 [DBG] 
 Buffer Header 
---------------- 
Length: 0 
Presentation Timestamp: -9223372036854775808 
Flags: 
MMAL_BUFFER_HEADER_FLAG_EOS 
MMAL_BUFFER_HEADER_FLAG_COMPLETEFRAME 
---------------- 

2020-10-09 08:33:01.473 -04:00 [DBG] In managed Output callback
2020-10-09 08:33:01.473 -04:00 [DBG] Getting data from buffer
2020-10-09 08:33:01.473 -04:00 [DBG] Attempting to process data.
2020-10-09 08:33:01.474 -04:00 [DBG] PostProcess stream length: 67729
2020-10-09 08:33:01.474 -04:00 [DBG] Applying image processing.
2020-10-09 08:33:01.474 -04:00 [DBG] Setting parameter MMAL_PARAMETER_CAPTURE
2020-10-09 08:33:01.476 -04:00 [INF] cell index 0```

Reply:
Now it works with raw or formatted contexts, but of course a formatted context adds conversion overhead. It's probably worth mentioning that in the wiki for performance-sensitive uses. I'll explain "local context" in a moment but:

640x480 raw input, formatted output:
```
2020-10-09 14:31:44.654 -04:00 [DBG] ApplyConvolution set local context: 0 ms
2020-10-09 14:31:44.765 -04:00 [DBG] ApplyConvolution processed 1024 cells: 109 ms
2020-10-09 14:31:44.860 -04:00 [DBG] ApplyConvolution copy/formatted the results: 95 ms
```

640 x 480 JPEG input, formatted output:
```
2020-10-09 14:35:30.124 -04:00 [DBG] ApplyConvolution set local context: 94 ms
2020-10-09 14:35:30.225 -04:00 [DBG] ApplyConvolution processed 1024 cells: 100 ms
2020-10-09 14:35:30.270 -04:00 [DBG] ApplyConvolution copy/formatted the results: 44 ms
```

It occurs to me that because of the `StoreFormat` output stage, requiring raw input may not have been so onerous in the first place, but the work is done so no point in going back now.

When the input context is not already raw, it uses the `Bitmap` class to decode the data, then converts the pixel format to one of the raw pixel formats via `Bitmap.Clone`. The data isn't _actually_ raw -- as we discussed earlier, raw means never-compressed -- but it's still a raw _format_ which lets other parts of the system that do require raw data to operate normally. That data gets copied into a new local context used for processing. (I got rid of the `ImageContext` formatting extension class and brought the necessary methods back into `ConvolutionBase`, as I think they're probably not generically useful.)

After processing, that local context data is saved back to the original context using the requested `StoreFormat`, or if formatting wasn't provided, if the source wasn't originally a raw format, it array-copies the local context data back to the source context. 

Whew. It sounds simple now, but it took some pacing around the house to get all of that straight!

For some reason, `ImageContext.Resolution` isn't set when using `TakePicture`, although `TakeRawPicture` does set it. For now I read it from `Bitmap` after it loads the data, so this isn't a problem, but perhaps something to look into.

Raw-based output is still blue-tinted, although using encoded input is not. Curious...

The `PostProcess` zero-length check described in the preceding message is still there, too, and I'd still like to understand whether it might be more appropriate in `PortCallbackHandler`.

New commit momentarily...

Reply:
To save you time, a recap of today's work:
* Convolution works with raw or formatted inputs
* Open question [here](https://github.com/techyian/MMALSharp/pull/175#discussion_r502351729) about how to deal with non-32-byte-aligned resolutions
* The blue-tint problem with raw input (encoding is wrong somewhere I suspect)
* Zero-length buffer/data check in `PostProcess` or `PortCallbackHandler`? Any side-effects?

Other items that might warrant later consideration:
* Should `StreamCaptureHandler` silently suppress exceptions? Re-throw if logging is off?
* `TakePicture` doesn't set `ImageContext.Resolution`

Reply:
Earlier I had overlooked that the padded buffer also changed the vertical resolution -- turns out that's padded to 16-byte values. That means almost all resolutions have a raw buffer size that is larger than the image resolution. For convolution, this only impacts `TakeRawPicture` since a formatted image from `TakePicture` is digested by `Bitmap` which produces a buffer matching the image resolution. 

I committed an update that just adds cell sizes for the padded buffer dimensions. I can't decide if it would be better for `ImageContext.Resolution` to match the camera resolution or the true padded buffer size. As far as I can tell, the only effect of matching camera resolution would be to ignore the padding, which is what you'd probably want anyway.

Reply:
> My TakeRawPicture test set the camera to 1296x972 as a side-effect of earlier motion-detection testing (it's a full-frame binning mode). To my surprise, a different resolution was reported in an exception in the debug log: Resolution 1312x976 has no recommended cell counts. After a bit of searching, it turns out that the driver bumps the buffer up to the nearest 32-byte value (RPi forum discussion):

I might need to do some investigation to check the camera's resolution is being configured correctly. The configuring of the camera component can be found in `MMALCameraComponent`. I'm not sure if the driver aligns on your behalf - I explicitly align the width and height to 32 and 16 respectively via the `Resolution` class and this library is influenced by Raspistill/Raspivid in terms of configuring the camera. 

> I'm wondering if you can think of where in the pipeline this might go wrong. (EDIT - I think it's the second part of the condition wrapping the output port callback, specifically (eos && !this.Trigger.Task.IsCompleted) ... I understand why it does that but I'm not sure how to address it.)

Well spotted, yes I think that's the problem. Could this be the solution:

```csharp
var eos = bufferImpl.AssertProperty(MMALBufferProperties.MMAL_BUFFER_HEADER_FLAG_FRAME_END) ||
                  bufferImpl.AssertProperty(MMALBufferProperties.MMAL_BUFFER_HEADER_FLAG_EOS) ||
                  this.ComponentReference.ForceStopProcessing;

if ((bufferImpl.CheckState() && bufferImpl.Length > 0 && !eos && !failed && !this.Trigger.Task.IsCompleted) || (eos && !this.Trigger.Task.IsCompleted && bufferImpl.Length > 0))
{
    this.CallbackHandler.Callback(bufferImpl);
}
```

Basically just making sure that the buffer length is always greater than 0 if we want to call the Callback Handler?

> For some reason, ImageContext.Resolution isn't set when using TakePicture, although TakeRawPicture does set it. For now I read it from Bitmap after it loads the data, so this isn't a problem, but perhaps something to look into.

Hmm, it should do. What is happening here is when `ConfigureOutputPort` is called against the encoder component (the output port being an instance of `StillPort`), the configure stage will check if the `MMALPortConfig` has a resolution set against it, and if not (which is the default), it will look [here](https://github.com/techyian/MMALSharp/blob/dev/src/MMALSharp/Ports/Outputs/StillPort.cs#L28) for the resolution to set it to. When your `ImageContext` is populated, the `Resolution` property is taken from the working port which in your case will be the encoder output port. If you are certain that it's not being retrieved correctly then please raise a ticket.

> Raw-based output is still blue-tinted, although using encoded input is not. Curious...

I have a feeling the raw output is being sent as BGR instead of RGB, I haven't had chance to look into this but it warrants a ticket raising rather than blocking this PR.

> Should StreamCaptureHandler silently suppress exceptions? Re-throw if logging is off?

Could you tell me where the exception is being suppressed and I'll review?

Reply:
You're right about the raw image and RGB vs BGR, I temporarily added something to the end of the cell processing method to swap the R/B bytes and that fixes the color problem. I don't know where to start looking for an explanation though. Hmm, I wonder if [this](https://www.raspberrypi.org/forums/viewtopic.php?t=262579) firmware issue is related? (Although I'm not clear about the context of that issue, it sounds a bit like it might be specific to HDMI.)

I don't think `Bitmap.Save` is at fault. Over in my little test program [here](https://github.com/MV10/pi-cam-test/blob/local_mmal_dev_pkg/Program.cs#L1168) I was using it to save raw images to JPG and had no issues. However, I think the raw files I was loading came from raspicam so maybe it does something differently. (I don't think I ever ran the library's `TakeRawPicture` method before I started working on the convolution changes.)

Reply:
Hi Ian, my browser's view into this PR hadn't updated with your comments until just now, after I'd already been playing around with the BGR/RGB issue. See comments above about the firmware issue which sounds related. I will also open a ticket about this -- I just now wrote a stand-alone test that does nothing but send `TakeRawPicture` to disk, then reads it back and does `Bitmap.Save` -- no other manipulation -- and the resulting JPEG has the colors reversed. Definitely not related to this PR.

> I might need to do some investigation to check the camera's resolution is being configured correctly. The configuring of the camera component can be found in MMALCameraComponent. I'm not sure if the driver aligns on your behalf - I explicitly align the width and height to 32 and 16 respectively via the Resolution class and this library is influenced by Raspistill/Raspivid in terms of configuring the camera.

I think this one is no longer a problem now that I've added the padded "resolutions" to the cell table, too. It's just the nature of the raw buffer that we have to work with it that way, I don't think there's benefit in messing with it further. About the only thing that _might_ make sense is to reflect the true resolution on `ImageContext` so that we don't waste time processing blank padding bytes, but honestly I think the overhead is immeasurably small.

>  Could this be the solution: Basically just making sure that the buffer length is always greater than 0 if we want to call the Callback Handler?

That does indeed fix the problem earlier in the pipeline! Committing that change now.

I will open a ticket about `TakePicture` not setting the resolution, I went so far as to put a `Console.WriteLine` as the first line of `ApplyConvolution` and it's definitely 0,0.

> Should StreamCaptureHandler silently suppress exceptions? Re-throw if logging is off?
> Could you tell me where the exception is being suppressed and I'll review?

https://github.com/MV10/MMALSharp/blob/motiondiff/src/MMALSharp.Processing/Handlers/StreamCaptureHandler.cs#L77

Reply:
Pushed another fix to copy the cloned bitmap resolution back to the original context (which would still be 0,0 when invoked from `TakePicture`) -- without this it wasn't possible to call multiple effects using `TakeRawPicture` (confusing, I know):

```csharp
imgCaptureHandler.Manipulate(context =>
{
    context.Apply(new SharpenProcessor());
    context.Apply(new GaussianProcessor(GaussianMatrix.Matrix5x5));
}, ImageFormat.Jpeg);
```

This was because after the first effects pass, `ImageContext` no longer holds formatted data, it is raw data. The second pass treats it as raw data, but raw data doesn't go through the clone process (to decode into a raw bitmap and derive the resolution). The code assumes raw context does have the correct resolution (`TakeRawPicture` _does_ set it), but in this scenario it hasn't been set because we didn't start from that method. Fun, eh?